src/v/cloud_topics/dl_stm/dl_stm_state.cc

```diff
@@ -0,0 +1,79 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "cloud_topics/dl_stm/dl_stm_state.h"
+
+#include "cloud_topics/dl_overlay.h"
+#include "model/fundamental.h"
+
+#include <algorithm>
+
+namespace experimental::cloud_topics {
+
+void dl_stm_state::push_overlay(dl_overlay overlay) {
+    vassert(
+      _current_version > dl_version(0),
+      "dl_stm_state must be initialized with a non-zero version before pushing "
+      "an overlay");
+
+    auto entry_it = std::find_if(
+      _overlays.begin(),
+      _overlays.end(),
+      [&overlay](const dl_overlay_entry& entry) {
+          return entry.overlay == overlay;
+      });
+    if (entry_it != _overlays.end()) {
```

Comment:
STM will not be applying the same command twice.

Reply:
I believe you are wrong about this one. Here is an example. [`throwing_kv`](https://github.com/redpanda-data/redpanda/blob/a96fcd1477453f8cc49a566287bf941755defbdc/src/v/raft/tests/stm_manager_test.cc#L20) has the following code inside `apply`

https://github.com/redpanda-data/redpanda/blob/a96fcd1477453f8cc49a566287bf941755defbdc/src/v/raft/tests/stm_manager_test.cc#L25-L30

so it will randomly succeed or throw. Here is log output from a [test](https://github.com/redpanda-data/redpanda/blob/a96fcd1477453f8cc49a566287bf941755defbdc/src/v/raft/tests/stm_manager_test.cc#L187) using it:

```cpp
# foreground exception caught and the stm manager switches to a background apply fiber
WARN  2024-10-29 11:55:14,539 [shard  0:main] raft - [group_id:123, {kafka/node_0/0}] state_machine_manager.cc:151 - [default][throwing_kv] error applying batch with base_offset: 7 - std::runtime_error (runtime error from throwing stm)

# the same batch (base_offset: 7) apply retried in background
WARN  2024-10-29 11:55:14,539 [shard  0:main] raft - [group_id:123, {kafka/node_0/0}] state_machine_manager.cc:151 - [background][throwing_kv] error applying batch with base_offset: 7 - std::runtime_error (runtime error from throwing stm)

# a very unlucky batch retried 6 times
WARN  2024-10-29 11:55:16,122 [shard  0:main] raft - [group_id:123, {kafka/node_0/0}] state_machine_manager.cc:151 - [background][throwing_kv] error applying batch with base_offset: 709 - std::runtime_error (runtime error from throwing stm)
WARN  2024-10-29 11:55:16,123 [shard  0:main] raft - [group_id:123, {kafka/node_0/0}] state_machine_manager.cc:151 - [background][throwing_kv] error applying batch with base_offset: 709 - std::runtime_error (runtime error from throwing stm)
WARN  2024-10-29 11:55:16,224 [shard  0:main] raft - [group_id:123, {kafka/node_0/0}] state_machine_manager.cc:151 - [background][throwing_kv] error applying batch with base_offset: 709 - std::runtime_error (runtime error from throwing stm)
WARN  2024-10-29 11:55:16,326 [shard  0:main] raft - [group_id:123, {kafka/node_0/0}] state_machine_manager.cc:151 - [background][throwing_kv] error applying batch with base_offset: 709 - std::runtime_error (runtime error from throwing stm)
WARN  2024-10-29 11:55:16,428 [shard  0:main] raft - [group_id:123, {kafka/node_0/0}] state_machine_manager.cc:151 - [background][throwing_kv] error applying batch with base_offset: 709 - std::runtime_error (runtime error from throwing stm)
WARN  2024-10-29 11:55:16,529 [shard  0:main] raft - [group_id:123, {kafka/node_0/0}] state_machine_manager.cc:151 - [background][throwing_kv] error applying batch with base_offset: 709 - std::runtime_error (runtime error from throwing stm)
WARN  2024-10-29 11:55:16,631 [shard  0:main] raft - [group_id:123, {kafka/node_0/0}] state_machine_manager.cc:151 - [background][throwing_kv] error applying batch with base_offset: 728 - std::runtime_error (runtime error from throwing stm)

# etc...
```
https://gist.github.com/nvartolomei/a560afa068c7bc8e013e48c2d6b8a325

Reply:
Correct. I think we're solving this by tracking insync offset in the archival stm. Why not using the same approach here instead of relying on more complex logic?